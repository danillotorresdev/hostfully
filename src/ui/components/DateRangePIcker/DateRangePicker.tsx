import { useRef, useState } from "react";
import * as S from "./DateRangePicker.styles";
import { Calendar } from "./Calendar/Calendar";
import { useClickoutside } from "@/utils/hooks/useClickOutside/useClickoutside";
import { Icon } from "@/ui/components/Icon/Icon";
import { Button } from "@/ui/components/Button/Button";

type DateRangePickerProps = {
  onSelect: (startDate: Date, endDate: Date) => void;
  defaultValue?: {
    startDate: Date;
    endDate: Date;
  };
};

export const DateRangePicker = ({
  onSelect,
  defaultValue,
}: DateRangePickerProps) => {
  const [startDate, setStartDate] = useState<Date | null>(
    defaultValue?.startDate ?? null,
  );
  const [endDate, setEndDate] = useState<Date | null>(
    defaultValue?.endDate ?? null,
  );
  const [showCalendar, setShowCalendar] = useState<boolean>(false);
  const [currentMonth, setCurrentMonth] = useState<number>(
    defaultValue?.startDate?.getMonth() ?? new Date().getMonth(),
  );
  const [currentYear, setCurrentYear] = useState<number>(
    defaultValue?.startDate?.getFullYear() ?? new Date().getFullYear(),
  );

  const calendarRef = useRef<HTMLDivElement>(null);

  useClickoutside({
    callback: () => setShowCalendar(false),
    ref: calendarRef,
  });

  const handleDayClick = (day: number, monthOffset: number) => {
    const clickedDate = new Date(currentYear, currentMonth + monthOffset, day);
    if (!startDate) {
      setStartDate(clickedDate);
      setEndDate(null);
    } else if (!endDate && clickedDate >= startDate) {
      setEndDate(clickedDate);
    } else {
      if (clickedDate.getMonth() === currentMonth + monthOffset) {
        setStartDate(clickedDate);
        setEndDate(null);
      }
    }
  };

  const renderDaysForMonth = (
    year: number,
    month: number,
    monthOffset: number,
  ): JSX.Element[] => {
    const days: JSX.Element[] = [];
    const totalDays = new Date(year, month + 1, 0).getDate();
    const firstDay = new Date(year, month, 1).getDay();
    const today = new Date();

    for (let i = 0; i < firstDay; i++) {
      days.push(<div key={`empty-${i}`} />);
    }

    for (let i = 1; i <= totalDays; i++) {
      const date = new Date(year, month, i);
      const isSelected =
        (startDate && date.toDateString() === startDate.toDateString()) ||
        (endDate && date.toDateString() === endDate.toDateString());
      const isInRange =
        startDate && endDate && date >= startDate && date <= endDate;
      const isDisabled = date < today;
      days.push(
        <S.Day
          key={i}
          isSelected={!!isSelected}
          isInRange={!!isInRange}
          isDisabled={isDisabled}
          onClick={() => !isDisabled && handleDayClick(i, monthOffset)}
        >
          {i}
        </S.Day>,
      );
    }

    return days;
  };

  const handlePrevMonth = () => {
    setCurrentMonth((prevMonth) => {
      if (prevMonth === 0) {
        setCurrentYear((prevYear) => prevYear - 1);
        return 11;
      }
      return prevMonth - 1;
    });
  };

  const handleNextMonth = () => {
    setCurrentMonth((prevMonth) => {
      if (prevMonth === 11) {
        setCurrentYear((prevYear) => prevYear + 1);
        return 0;
      }
      return prevMonth + 1;
    });
  };

  const toggleCalendar = () => {
    setShowCalendar(!showCalendar);
  };

  return (
    <S.Container>
      <S.DateInputLabel>Date</S.DateInputLabel>
      <S.DateInputContainer onClick={toggleCalendar}>
        <S.DepartureReturnDateWrapper>
          <S.DepartureReturnContent>
            <Icon name="calendar" />
            {startDate ? startDate.toLocaleDateString() : "Departure"}
          </S.DepartureReturnContent>
          <S.DepartureReturnContent>
            <Icon name="calendar" />
            {endDate ? endDate.toLocaleDateString() : "Return"}
          </S.DepartureReturnContent>
        </S.DepartureReturnDateWrapper>
      </S.DateInputContainer>

      {showCalendar && (
        <S.CalendarWrapper ref={calendarRef}>
          <Calendar
            currentMonth={currentMonth}
            currentYear={currentYear}
            handleNextMonth={handleNextMonth}
            handlePrevMonth={handlePrevMonth}
            renderDaysForMonth={renderDaysForMonth}
            className="calendar"
          />

          <S.CalendarFooter>
            <Button
              type="button"
              size="sm"
              onClick={() => {
                if (startDate && endDate) {
                  onSelect(startDate, endDate);
                  setShowCalendar(false);
                }
              }}
            >
              Select
            </Button>
          </S.CalendarFooter>
        </S.CalendarWrapper>
      )}
    </S.Container>
  );
};
