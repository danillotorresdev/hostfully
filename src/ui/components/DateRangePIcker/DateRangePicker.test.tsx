import { DateRangePicker } from "./DateRangePicker";
import { expect, it, vi } from "vitest";
import { describe } from "node:test";
import {
  render,
  screen,
  waitFor,
  fireEvent,
} from "@/utils/settings/tests/utilities";

describe("DateRangePicker", () => {
  it("should render with default values", () => {
    const defaultValues = {
      startDate: new Date("2024-09-11T03:00:00.000Z"),
      endDate: new Date("2024-11-11T03:00:00.000Z"),
    };

    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );

    expect(screen.getByText("9/11/2024")).toBeDefined();
    expect(screen.getByText("11/11/2024")).toBeDefined();
  });

  it("should call onSelect when a date is selected", async () => {
    const defaultValues = {
      startDate: new Date("2024-09-11T03:00:00.000Z"),
      endDate: new Date("2024-11-11T03:00:00.000Z"),
    };
    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const getStartDate = screen.getByText("15");
    const getEndDate = screen.getByText("20");
    fireEvent(getStartDate, new MouseEvent("click", { bubbles: true }));
    fireEvent(getEndDate, new MouseEvent("click", { bubbles: true }));

    const selectButton = screen.getByText("Select");
    fireEvent(selectButton, new MouseEvent("click", { bubbles: true }));

    await waitFor(() => {
      expect(onSelectMock).toHaveBeenCalled();
    });
  });

  it("should navigate to the next month", () => {
    const defaultValues = {
      startDate: new Date("2024-09-11T03:00:00.000Z"),
      endDate: new Date("2024-11-11T03:00:00.000Z"),
    };
    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const nextMonthButton = screen.getByLabelText("next month");
    fireEvent.click(nextMonthButton);

    expect(screen.getByText("October 2024")).toBeDefined();
  });
  it("should navigate to the previous month", () => {
    const defaultValues = {
      startDate: new Date("2024-09-11T03:00:00.000Z"),
      endDate: new Date("2024-11-11T03:00:00.000Z"),
    };
    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const prevMonthButton = screen.getByLabelText("previous month");
    fireEvent.click(prevMonthButton);

    expect(screen.getByText("August 2024")).toBeDefined();
  });

  it("should navigate to the next year when the next month is December", () => {
    const defaultValues = {
      startDate: new Date("2024-12-11T03:00:00.000Z"),
      endDate: new Date("2024-12-11T03:00:00.000Z"),
    };
    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const nextMonthButton = screen.getByLabelText("next month");
    fireEvent.click(nextMonthButton);

    expect(screen.getByText("January 2025")).toBeDefined();
  });

  it("should navigate to the previous year when the previous month is January", () => {
    const defaultValues = {
      startDate: new Date("2024-01-11T03:00:00.000Z"),
      endDate: new Date("2024-01-11T03:00:00.000Z"),
    };
    const onSelectMock = vi.fn();

    render(
      <DateRangePicker onSelect={onSelectMock} defaultValue={defaultValues} />,
    );
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const prevMonthButton = screen.getByLabelText("previous month");
    fireEvent.click(prevMonthButton);

    expect(screen.getByText("December 2023")).toBeDefined();
  });

  it("should click in a day and select the start date", () => {
    const onSelectMock = vi.fn();

    render(<DateRangePicker onSelect={onSelectMock} />);
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const getStartDate = screen.getByText("15");
    fireEvent.click(getStartDate);

    expect(screen.getByText("15")).toBeDefined();
  });

  it("should close the calendar when clicking outside", () => {
    const onSelectMock = vi.fn();

    render(<DateRangePicker onSelect={onSelectMock} />);

    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const outsideElement = document.body;
    fireEvent.mouseDown(outsideElement);

    expect(screen.queryByTestId("leftCalendar")).toBeNull();
    expect(screen.queryByTestId("rightCalendar")).toBeNull();
  });
  it("should select the start date if the start date has not been previously selected", () => {
    const onSelectMock = vi.fn();

    vi.useFakeTimers();
    vi.setSystemTime(new Date("2024-09-14"));

    render(<DateRangePicker onSelect={onSelectMock} />);
    const departureLabel = screen.getByLabelText("date");

    fireEvent.click(departureLabel);

    const getStartDate = screen.getByText("15");
    fireEvent.click(getStartDate);

    expect(screen.getByText("15")).toBeDefined();
  });
});
