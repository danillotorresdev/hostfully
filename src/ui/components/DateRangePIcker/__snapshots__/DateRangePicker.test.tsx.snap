// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render with default values 1`] = `
<DocumentFragment>
  <div
    class="mocked-styled-0 cx3igb0"
  >
    <label
      class="mocked-styled-2 d1m4j7uz"
    >
      Date
    </label>
    <div
      aria-label="date"
      class="mocked-styled-1 d6fdk0w"
    >
      <div
        class="mocked-styled-5 dv2dzrz"
      >
        <div
          class="mocked-styled-6 d1xj2x14"
        >
          <svg
            class="svg svg-calendar"
            data-testid="icon-svg"
            style="width: 24px; height: 24px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#calendar"
            />
          </svg>
          9/11/2024
        </div>
        <div
          class="mocked-styled-6 d1xj2x14"
        >
          <svg
            class="svg svg-calendar"
            data-testid="icon-svg"
            style="width: 24px; height: 24px;"
            viewBox="0 0 1024 1024"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="/assets/sprite.svg#calendar"
            />
          </svg>
          11/11/2024
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
